### ISSUE TITLE
CLI option for start-storybook to not open the browser

### ISSUE THREAD

**Issue by simbo** [storybookjs/storybook#6201]
**Is your feature request related to a problem? Please describe.**
I restart `start-storybook` from time to time for different reasons from changed config to working on other things. Nevertheless, I do not close my storybook browser tab. And as `start-storybook` opens a new browser tab every time it starts, my tab bar gets filled up with multiple storybook tabs between other tabs. It's annoying and leads to confusion and tab cleanup time...

I also noticed my colleagues doing the same things resulting in the same problems... 
And I was a bit surprised, that I couldn't find an option to not open the browser...

I know, there  is `--ci` to skip interactive prompts and not open the browser, but it also skips interactive prompts. Although I didn't see any interactive prompts yet, using ci mode for daily usage just doesn't feel right... 

**Describe the solution you'd like**

`start-server --no-open` should build and serve as usual but not open the browser window.

**Describe alternatives you've considered**

Opening the browser with `start-storybook` should be disabled by default and can be enabled by using `--open`.

Webpack supports the `--open`/ `--no-open` flags which are also mirrored by other CLIs which use webpack internally. 
angular cli  and vue cli offer the same flags and - like webpack-dev-server - don't open the browser by default.

From my personal perspective, opening the browser by default is an opinionated bad-practice.
But I know, there are different opinions about this out there... :)

**Are you able to assist bring the feature to reality?**
Yes, I could dig into this and make a PR.

**Additional context**
Anybody want a screenshot of my browser tab bar filled with storybook tabs? ;)

**Comment by shilman** [comment]
Hi @simbo. AFAIK the `--ci` option does exactly that. Are you proposing renaming it, and making it true by default?

**Comment by simbo** [comment]
It seems so... :)

**Comment by shilman** [comment]
**Alias.** I'm fine with adding an alias to make the feature more discoverable. We could also make it a separate flag if at some point there are more things that we don't want to do when we're in CI aside from just not opening the browser.

**Default value.** As for changing the defaults, I'm not opposed to it, though it would be a breaking change so we'd want to do it in 6.0. I'm posting this issue in the `#maintenance` Discord so hopefully we can get some consensus. If so, it's a trivial change but it will probably take awhile to release.

**Comment by bebbi** [comment]
It may not be the same. I think CRA handles this in a graceful way and could be adopted. Key differences are:

- target the tab that's already open instead of opening a new one every time.
- set preferred browser or do not open -> .env file
- favor Chrome over default browsers (e.g. for people that browse with a different browser than they develop, of which there seem many - Chrome is still a safer bet for the majority).

Relevant [CRA documentation](https://facebook.github.io/create-react-app/docs/advanced-configuration)

For me, start-storybook opens a tab in a backgrounded window of a different browser everytime I restart it.

**Comment by Hypnosphi** [comment]
> it also skips interactive prompts

One example of such prompt is about running Storybook an alternative port when the requested one is busy. That can be quite useful sometimes. So I think, `--no-open` shouldn't be an alias for `--ci`. It should only disable the opening

**Comment by Hypnosphi** [comment]
> favor Chrome over default browsers

Yeah, that's why I didn't copy that approach as-is

**Comment by backbone87** [comment]
personally i would prefer `--no-open` to be the default

**Comment by shilman** [comment]
Does anybody prefer the current behavior? If not, I think we can make the change

**Comment by Hypnosphi** [comment]
I do

**Comment by backbone87** [comment]
maybe we just provide both cli flags: `--option` and `--no-open` and allow to configure this once `storybook.config.js` landed.

**Comment by kutenai** [comment]
I would prefer a --no-open flag as well. Do everything that is currently default, just skip opening a new browser. I *VERY* often have an existing window open but must close and re-start the storybook process (due to configuration changes, etc), so I cannot rely on the auto-restart. In those cases, I just don't want a new browser.

**Comment by shilman** [comment]
@kutenai Can you just run with `--ci`? The discussion here is whether that behavior should be the default.

**Comment by kutenai** [comment]
It appears to work fine for me. Maybe the option is just misleading, as it kind of implies that it does "more" than I want, but in practice, it seems to work fine.

**Comment by jessepinho** [comment]
FWIW, create-react-app lets you set a `BROWSER` env variable to choose which browser to open, or `none` to prevent any browser from opening. Could be worth looking into

**Comment by Hypnosphi** [comment]
> The discussion here is whether that behavior should be the default.

You actually liked my comment that says that it should be a separate option

**Comment by shilman** [comment]
> You actually liked my comment that says that it should be a separate option

Yeah I'm absolutely on board with making `--no-open` a separate option. I also think it should be the default and that's what I'm hoping to get consensus on for 6.0

**Comment by SeedyROM** [comment]
The `--no-open` being default seems like a sane default to have, honestly opening a browser window by default seems like it's mostly for flash in a conference talk and does nothing but annoy developers who actually reload this command quite often.

**For now:**
Is there actually a solution to not opening a browser window?

**Comment by jaredh159** [comment]
I would also like the default to be not to open the browser, and fwiw my first guess was that I could add `BROWSER=none` like cra -- so it might be worth adopting that method for consistency.  I might not be the only one who would guess that before looking at docs.

**Comment by JonKrone** [comment]
My understanding of react-script's `start` behavior is that by default it opens a tab and if there is one already open, it re-uses that tab.

I love that behavior. If I've been working on a project through code changes and inevitable app/webpack crashes, I don't have to worry about bloating my windows with new tabs of the same app or flipping a flag to open a new tab.

We should also be careful of having two commands to run during normal dev workflows: `start-storybook --open` on the initial start and then `start-storybook [--no-open]` on subsequent commands. If tab re-use is possible, this is not a problem as my workflow would then just always be `start-storybook --open`.

I would prefer Storybook's default be to open a tab but am fine if y'all decide not to. I would love to see the tab re-use behavior come through though.

**Comment by shilman** [comment]
@JonKrone Unfortunately it only reuses the existing tab on HMR as far as I know. If I re-run the CLI it always opens a new tab for me (on Chrome/OSX at least).

**Comment by SebbeJohansson** [comment]
I can not get --ci to work. Still opens a new tab in chrome each time I start it. I'm on 5.2.1 with vue.

**Comment by shilman** [comment]
@SebbeJohansson can you double check? working for me

**Comment by SebbeJohansson** [comment]
@shilman I just did with a completely new setup just to make sure.
\My-Storybook>npm run storybook --ci
Results in a new tab in my primary browser.

**Comment by Hypnosphi** [comment]
@SebbeJohansson when using `npm run`, you have to prepend `--` to all the arguments that you want to pass to the underlying command:
```
npm run storybook -- --ci
```

**Comment by vandijkstef** [comment]
As long as restarting isn't properly targeting the same browser window/tab (and/or reloading it), opening a new window by default is just frustrating for me. I think this is an anti-pattern a lot of packages are using to be *fancy or something*. We're all capable of opening a browser by the time we know how to handle npm.

After building it shows a nice box with 2 url's I can ctrl-click on, and most of us probably use `start-storybook` while developing and restarting **a lot**.

TLDR: Upvoting `--ci` as default and opening in browser by flag

**Comment by Hypnosphi** [comment]
> restarting a *lot*

Why do you need that? `start-storybook` should pick your changes. If it doesn't, it sounds like an issue by itself.

> We're all capable of opening a browser

Of course you know *how* to open the browser, the problem is that you don't know *when*. In large projects, first build may take quite a while

**Comment by vandijkstef** [comment]
> Why do you need that? `start-storybook` should pick your changes. If it doesn't, it sounds like an issue by itself.

It should, and does on most changes, but not all changes. I'm currently setting up a new Storybook and fiddling around with it, so yes I ended up with 10 tabs in a few minutes.

> Of course you know _how_ to open the browser, the problem is that you don't know _when_. In large projects, first build may take quite a while, and just sitting and waiting for the box in the console may be not an option

I could still open the browser, it'll just tell me 'Connection refused'. I think it's more annoying that after a minute or 2 suddenly a browser is popping up while I'm doing something else.
There could be a comment in the box there is a `--automagically-open-in-browser` (working title) flag you can enable.

Note I'm seeing this pattern across the industry, and where used, issues, & (stackoverflow) questions pop up everywhere how to disable it. I heavily dislike the pattern **as a default**, especially since targeting the same tab isn't solid technology. Meanwhile, once opened, a disconnect to the 'server' could start polling and reload once it comes back up again.

**Comment by shilman** [comment]
Add a message to CLI output notifying the user of `--ci` option if they don't want it to show cc @yannbf

**Comment by cloud-walker** [comment]
Where can we see these improvements?

**Comment by shilman** [comment]
Available in 6.0-beta. https://github.com/storybookjs/storybook/pull/10329

**Comment by MickL** [comment]
I am not sure how this is related to continuous integration. In fact I am developing and it is annoying that storybook opens the browser on start. I am not doing ci so imo the naming is not correct. I would like to have a `--open` false option, too. Or even go for not open storybook by default (by comparison Angular CLI does not open by default, too) and then make the option `--open true` available.

**Comment by balibebas** [comment]
> Proposal: Instead of doing all this, can we just add a message to CLI output notifying the user of `--ci` option if they don't want it to open the browser? The current behavior is much better thanks to @yannbf 's improvements

I landed here trying to stop the browser window from opening. If it could detect an existing open tab and reuse it I think 99% of people would be fine with that. But the concern with this issue (which has 47 upvotes at current) is that they simply don't want to have the browser window open and `--ci` is a misnomer.

**Comment by 100terres** [comment]
Hi everyone! I've took a look at the code of the library used to open the browser ([beter-opn](https://github.com/ExiaSR/better-opn/tree/master)) and I've found these lines

https://github.com/ExiaSR/better-opn/blob/37065aa27fa1e5ad8c9f78a993d22b8ec3d9db23/src/index.js#L115-L117

```tsx
case Actions.NONE:
  // Special case: BROWSER="none" will prevent opening completely.
  return false;
```

I've tried to add the environment variable `BROWSER=none` to my script and it seems to be working as expected!

Here's what I have in my `package.json` in the `scripts` section. Using the latest version of storybook.

```
"storybook": "BROWSER=none start-storybook",
```

A simple mention of the environement variable in the documentation could do the job. No code would need to be changed for it to work :tada: :slightly_smiling_face:

**Comment by shilman** [comment]
cc @jonniebigodes ☝️

**Comment by shilman** [comment]
Proposal: create a `--open` flag (which is true by default so this wouldn't be a breaking change) and then `--no-open` will set it to false. If open is false, then we simply don't call `better-opn`. Should be a small change! 💪

**Comment by roberto-red** [comment]
Just a heads-up: it's probably undocumented, but setting a `BROWSER` environment variable to `none` on a `.env` file (eg. `.env.local`, to keep the change to yourself) results in `start-storybook` not opening a browser tab. It's the same behaviour `create-react-app` has. I just discovered this on Storybook 6.3.4.

`.env.local`:
```
BROWSER=none
```

**Comment by niieani** [comment]
this auto-start browser leads to unfortunate bugs like: https://github.com/storybookjs/storybook/issues/34631
It would be great if we could at least configure the default in the config file
